src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java

```diff
@@ -0,0 +1,243 @@
+    private final boolean macroFirst;
+
+    // Fields in config.jelly must match the parameter names in the "DataBoundConstructor"
+    @DataBoundConstructor
+    public BuildNameUpdater(BuildNameFromFile nameFromFile, BuildNameFromMacro nameFromMacro) {
+        if (nameFromFile != null) {
+            this.buildName = nameFromFile.buildName;
+            this.fromFile = true;
+        }
+        else {
+            this.buildName = "";
+            this.fromFile = false;
+        }
+
+        if (nameFromMacro != null) {
+            this.macroTemplate = nameFromMacro.macroTemplate;
+            this.fromMacro = true;
+            this.macroFirst = nameFromMacro.macroFirst;
+        }
+        else {
+            this.macroTemplate = "";
+            this.fromMacro = false;
+            this.macroFirst = false;
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public boolean getFromFile() {
+        return fromFile;
+    }
+
+    @SuppressWarnings("unused")
+    public boolean getMacroFirst() {
+        return macroFirst;
+    }
+
+    @SuppressWarnings("unused")
+    public boolean getFromMacro() {
+        return fromMacro;
+    }
+
+    @SuppressWarnings("unused")
+    public String getBuildName() {
+        return buildName;
+    }
+
+    @SuppressWarnings("unused")
+    public String getMacroTemplate() {
+        return macroTemplate;
+    }
+
+    @Override
+    public boolean perform(AbstractBuild build, Launcher launcher, BuildListener listener) {
+        String buildNameToSet = "";
+
+        if (fromFile){
+            buildNameToSet = readFromFile(build, listener, buildName);
+        }
+
+        if(fromMacro){
+            String evaluatedMacro = getFromMacro(build, listener, macroTemplate);
+
+            listener.getLogger().println("Evaluated macro: '" + evaluatedMacro + "'");
+
+            buildNameToSet = macroFirst ? evaluatedMacro + buildNameToSet : buildNameToSet + evaluatedMacro;
+        }
+
+        if (buildNameToSet != null && !buildNameToSet.isEmpty()) {
+            setDisplayName(build, listener, buildNameToSet);
+        }
+
+        return true;
+    }
+
+    private void setDisplayName(AbstractBuild build, BuildListener listener, String result) {
+        listener.getLogger().println("Setting build name to '" + result + "'");
+        if (result.isEmpty()) {
+            listener.getLogger().println("Build name is empty, nothing to set.");
+            return;
+        }
+        try {
+            build.setDisplayName(result);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private String getFromMacro(AbstractBuild build, BuildListener listener, String macro){
+        String result = null;
+        try {
+            result = TokenMacro.expandAll(build, listener, macro);
+        } catch (MacroEvaluationException e) {
+            listener.getLogger().println("Failed to evaluate macro '" + macro + "'");
+            e.printStackTrace();
```

Comment:
This Stacktrace goes to Jenkins STDERR => No sense to do it. Just use java.util.logging.Logger.

Example: https://github.com/jenkinsci/sshd-module/blob/c18c8c1d7c30508d581ca004d7b09b38f5d5a289/src/main/java/org/jenkinsci/main/modules/sshd/ItemListenerImpl.java